src/cargo/util/template.rs

```diff
@@ -0,0 +1,231 @@
+    }
+    Ok(())
+}
+
+/// Trait to hold information required for rendering templated files.
+pub trait TemplateFile {
+    /// Path of the template output for the file being written.
+    fn path(&self) -> &Path;
+
+    /// Return the template string.
+    fn template(&self) -> CargoResult<String>;
+}
+
+/// TemplateFile based on an input file.
+pub struct InputFileTemplateFile {
+    input_path: PathBuf,
+    output_path: PathBuf,
+}
+
+impl TemplateFile for InputFileTemplateFile {
+    fn path(&self) -> &Path {
+        &self.output_path
+    }
+
+    fn template(&self) -> CargoResult<String> {
+        let mut template_str = String::new();
+        let mut entry_file = try!(File::open(&self.input_path).chain_error(|| {
+            human(format!("Failed to open file for templating: {}", self.input_path.display()))
+        }));
+        try!(entry_file.read_to_string(&mut template_str).chain_error(|| {
+            human(format!("Failed to read file for templating: {}", self.input_path.display()))
+        }));
+        Ok(template_str)
+    }
+}
+
+impl InputFileTemplateFile {
+    pub fn new(input_path: PathBuf, output_path: PathBuf) -> InputFileTemplateFile {
+        InputFileTemplateFile {
+            input_path: input_path,
+            output_path: output_path
+        }
+    }
+}
+
+/// An in memory template file for --bin or --lib.
+pub struct InMemoryTemplateFile {
+    template_str: String,
+    output_path: PathBuf,
+}
+
+impl TemplateFile for InMemoryTemplateFile {
+    fn path(&self) -> &Path {
+        &self.output_path
+    }
+
+    fn template(&self) -> CargoResult<String> {
+        Ok(self.template_str.clone())
+    }
+}
+
+impl InMemoryTemplateFile {
+    pub fn new(output_path: PathBuf, template_str: String) -> InMemoryTemplateFile {
+        InMemoryTemplateFile {
+            template_str: template_str,
+            output_path: output_path
+        }
+    }
+}
+
+pub enum TemplateDirectory{
+    Temp(TempDir),
+    Normal(PathBuf),
+}
+
+impl TemplateDirectory {
+    pub fn path(&self) -> &Path {
+        match *self {
+            TemplateDirectory::Temp(ref tempdir) => tempdir.path(),
+            TemplateDirectory::Normal(ref path) => path.as_path()
+        }
+    }
+}
+
+/// A listing of all the files that are part of the template.
+pub struct TemplateSet {
+    pub template_dir: Option<TemplateDirectory>,
+    pub template_files: Vec<Box<TemplateFile>>
+}
+
+// The type of template we will use.
+#[derive(Debug, Eq, PartialEq)]
+pub enum TemplateType<'a>  {
+    GitRepo(&'a str),
+    LocalDir(&'a str),
+    Builtin
+}
+
+/// Given a repository string and subdir, determine if this is a git repository, local file, or a
+/// built in template.
+pub fn get_template_type<'a>(repo: Option<&'a str>,
+                             subdir: Option<&'a str>) -> CargoResult<TemplateType<'a>> {
+    match (repo, subdir) {
+        (Some(repo_url), _) if repo_url.starts_with("http://")  ||
```

Comment:
In `tests/new.rs`: `git_template` if template treats `file://` as a URL then the following results in an error:
```rust
    assert_that(cargo_process("new").arg("--template").arg(git_project.url().as_str())
                                    .arg("foo")                                 
                                    .env("USER", "foo"),                        
                execs().with_status(0)); 
```

Error:
```
--- stderr
     Cloning file:///Users/ehiggs/src/cargo/target/cit/t1/template1
error: Failed to create project `foo` at `/Users/ehiggs/src/cargo/target/cit/t1/foo`

Caused by:
  Failed to read file for templating: /var/folders/nn/2zxw6dwn17l_vk2xzpj9s_8m0000gp/T/foo.wRUS45BLqSY6/objects/pack/pack-68af5d9cffc2ad3bdb15d6f776a4878c7f8a3670.idx
```
It went away when I treated `file://` as a file.

It's not just the test infrastructure. It also failed when I run it from the cmdline:

```
$ ./target/debug/cargo new --template file://$HOME/src/rust-cmdline-skeleton foo
     Cloning file:///Users/ehiggs/src/rust-cmdline-skeleton
error: Failed to create project `foo` at `/Users/ehiggs/src/cargo/foo`

Caused by:
  Failed to read file for templating: /var/folders/nn/2zxw6dwn17l_vk2xzpj9s_8m0000gp/T/foo.b3gbzHi1Ekl5/objects/pack/pack-47713ed0783c0ed9f74a83830bd6f43b7c8c0146.idx

To learn more, run the command again with --verbose.
```